flowjax/train/variational_fit.py

```diff
@@ -0,0 +1,89 @@
+from typing import Optional, Callable
+
+from jax import random
+import optax
+import equinox as eqx
+from tqdm import tqdm
+
+from flowjax.utils import Array
+from flowjax.distributions import Distribution
+
+# Variational targets accept an array of proposed samples and return an array of (usually)
+# the unormalised log posterior probability of the samples.
+VariationalTarget = Callable[[Array], Array]
+
+# Variational losses are functions that take a distribution, a target 
+# callable and a random key, and returns a scalar loss.
+VariationalLoss = Callable[
+    [Distribution, VariationalTarget, random.KeyArray], 
+    float
+]
+
+@eqx.filter_jit
+def elbo_loss(dist: Distribution, target: VariationalTarget, key: random.KeyArray, elbo_samples: int = 500):
+    samples = dist.sample(key, sample_shape=(elbo_samples,))
+    approx_density = dist.log_prob(samples).reshape(-1)
+    target_density = target(samples).reshape(-1)
+    losses = approx_density - target_density
+    return losses.mean()
+
+def variational_fit(
+    key: random.KeyArray,
+    dist: Distribution,
+    target: VariationalTarget,
+    loss_fcn: VariationalLoss = elbo_loss,
+    num_epochs: int = 100,
+    optimizer: Optional[optax.GradientTransformation] = None,
+    show_progress: bool = True,
+    recorder: Optional[Callable] = None,
+):
+    """
+    Train a distribution (e.g. a flow) by variational inference.
+    By default this does nothing fancy, just runs SGD steps on the ELBO loss for a fixed number of epochs.
+    Args:
+        key (KeyArray): Jax PRNGKey.
+        dist (Distribution): Distribution object, trainable parameters are found using equinox.is_inexact_array.
+        target (VariationalTarget): The target (usually) unormalized log posterior.
+        loss_fcn (VariationalLoss, optional): Loss function. Defaults to elbo_loss.
+        num_epochs (int, optional): The number of training steps to run. Defaults to 100.
+        optimizer (Optional[optax.Optimizer], optional): An optax optimizer (optimizers are implemented as GradientTransformation objects). 
+                                                         Defaults to an adam optimizer with learning rate 5e-4.
+        show_progress (bool, optional): Whether to show progress bar. Defaults to True.
+    """
+    @eqx.filter_jit
+    def step(dist, target, key, optimizer, opt_state):
+        loss_val, grads = eqx.filter_value_and_grad(loss_fcn)(dist, target, key)
+        updates, opt_state = optimizer.update(grads, opt_state)
+        dist = eqx.apply_updates(dist, updates)
+        return dist, opt_state, loss_val
+
+    # Set up a default optimizer if None is provided
+    if optimizer is None:
```

Comment:
I think that being able to pass an optimizer is reasonable, although it does lack consistency with `train_flow`. Especially with learning rate being such an important parameter, it seems wrong to have a hard-coded default, that you can only override by passing an optimizer.

My initial thoughts are we have three options:
1. Mirror the `train_flow` arguments, but add another optional argument for the optimizer. If the opimizer is passed, then `learning_rate` and `clip_norm` are ignored (and document this behaviour). We could modify `train_flow` similarly to allow passing of an optimizer.
2.  Set learning_rate=None, clip_norm=None and optimizer=None in the signature. Explicitly check that if optimizer is passed, both learning_rate and clip_norm are None. Then if an optimizer isn't passed, create one with learning_rate and clip_norm (setting them to default values if None).
3. Change `train_flow` and `variational_fit` to require an optimizer, with no default, and force the user to provide an optimizer.

I think I would lean towards option 2, but let me know what you think or if you have any other ideas

Reply:
One advantage of 3, is that we could also return the optimizer from the training scripts, and reuse them in further training, but it would make things a bit more verbose.

Reply:
I think in the interest of flexibility I think it would be good to support passing custom optimizers - I can imagine this would be something people want to play with.
On the other hand, I think you are right that the importance of the learning rate suggests it should be easier to access.

Honestly, on reflection I will just mirror the `train_flow` arguments, then we can solve the thoughts around a custom optimiser when it is actually needed.

Reply:
that change is in place now